### PR TITLE
JAMES-2554 Add test on Cassandra MailRepository storing implementation

### PR DESCRIPTION
**Pull request by aduprat**
The aim of those tests is to prouve the order of storing data:
MimeMessage -> message blob parts -> keys -> mail repository size

### CODE REVIEW

**Comment by aduprat** [comment]
```
+ docker run -a stdout -a stderr --tmpfs=/tmp:rw,exec,size=1g --tmpfs=/james-project:rw,size=4g -v /var/run/docker.sock:/var/run/docker.sock --volumes-from=keys -v maven-cache-volume:/root/.m2 --volumes-from=git-publish-james-2554-b2b3eb49-4338-4b12-8067-6eb30ae722fa --volume=/jpa/destination --volume=/cassandra/destination --volume=/spring/destination --name=james-parent-test-james-2554-b2b3eb49-4338-4b12-8067-6eb30ae722fa --env JAMES_PARTIAL_BUILD_BASE_BRANCH=origin/master --env MVN_ADDITIONAL_ARG_LINE=-Ppartial-build james-parent-compile-james-2554-b2b3eb49-4338-4b12-8067-6eb30ae722fa MERGE-james-2554-b2b3eb49-4338-4b12-8067-6eb30ae722fa
docker: Error response from daemon: No such container: keys.
See 'docker run --help'.
```
test this please

**Comment by aduprat** [comment]
> I would write that as 3 @nested tests because they are related
Not really

:+1:

File: server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+                return CompletableFuture.supplyAsync(() -> {
+                    throw new RuntimeException("Expected failure while storing keys");
+                });
+            }
+        }
+
+        @Test
+        void sizeShouldNotBeIncreasedWhenStoringKeysHasFailed() throws Exception {
+            MailKey mailKey = new MailKey("mymail");
+            List<MailAddress> recipients = ImmutableList
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java]: You can use LailAddressFixture to make this test lighter

**Comment by chibenwa** [comment]
test this pleasse

**Comment by chibenwa** [comment]
```
[ec02415c6925775c0e57e3009d990ab6462ac549] [ERROR] org.apache.james.mailrepository.cassandra.CassandraMailRepositoryWithFakeImplementationsTest  Time elapsed: 0.069 s  <<< ERROR!
[ec02415c6925775c0e57e3009d990ab6462ac549] java.util.concurrent.CompletionException: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (no host was tried)
[ec02415c6925775c0e57e3009d990ab6462ac549] Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (no host was tried)
[ec02415c6925775c0e57e3009d990ab6462ac549] 
[ec02415c6925775c0e57e3009d990ab6462ac549] [INFO] 
[ec02415c6925775c0e57e3009d990ab6462ac549] [INFO] Results:
[ec02415c6925775c0e57e3009d990ab6462ac549] [INFO] 
[ec02415c6925775c0e57e3009d990ab6462ac549] [ERROR] Errors: 
[ec02415c6925775c0e57e3009d990ab6462ac549] [ERROR]   CassandraMailRepositoryWithFakeImplementationsTest » Completion com.datastax.d...
```

test this please

**Comment by chibenwa** [comment]
Same error again

**Comment by aduprat** [comment]
> Same error again

Fixed, so my conclusion is that `@RegisterExtension` is not compatible with `@Nested`:
-> the lifecycle methods are called for each `@Nested` classes, this means that we are calling `@AfterAll` n times.

File: server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+            assertThatThrownBy(() -> cassandraMailRepository.store(mail))
+                    .isInstanceOf(RuntimeException.class)
+                    .hasMessage("java.lang.RuntimeException: Expected failure while storing mail parts");
+
+            assertThat(keysDAO.list(URL).join()).isEmpty();
+        }
+
+        @Test
+        void mimeMessageShouldBeStoredWhenStoringMailPartsHasFailed(CassandraCluster cassandra) throws Exception {
```
Comment (rouazana) [inline server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java]: why?

File: server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+            assertThatThrownBy(() -> cassandraMailRepository.store(mail))
+                    .isInstanceOf(RuntimeException.class)
+                    .hasMessage("java.lang.RuntimeException: Expected failure while storing keys");
+
+            assertThat(countDAO.getCount(URL).join()).isEqualTo(0);
+        }
+
+        @Test
+        void mimeMessageShouldBeStoredWhenStoringKeysHasFailed(CassandraCluster cassandra) throws Exception {
```
Comment (rouazana) [inline server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java]: why?

File: server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                    .isInstanceOf(RuntimeException.class)
+                    .hasMessage("java.lang.RuntimeException: Expected failure while storing keys");
+
+            ResultSet resultSet = cassandra.getConf().execute(select()
+                    .from(BlobTable.TABLE_NAME));
+            assertThat(resultSet.all()).hasSize(2);
+        }
+
+        @Test
+        void mailPartsShouldBeStoredWhenStoringKeysHasFailed(CassandraCluster cassandra) throws Exception {
```
Comment (rouazana) [inline server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java]: why?

File: server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java (hunk 2 above)
Comment (chibenwa) [reply server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java]: IMO this test is not here to enforce a behaviour but demonstate a behaviour.

This should maybe reflect in the test naming:

```
x Is y When z

Here: mimeMessageIsStoredWhenStoringMailPartsHasFailed
```

**Comment by chibenwa** [comment]
```
[46c48e10db27ecc48e62c595268f1a16a2226595] [ERROR] Errors: 
[46c48e10db27ecc48e62c595268f1a16a2226595] [ERROR]   SpamAssassinContract.deletingSpamMessageShouldNotImpactSpamAssassinState:540 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**